### PR TITLE
audience applause meter

### PR DESCRIPTION
**Pull request by amvmoody**
# QA

## Info
- recommended to be tested with a group.

## Setup
 - Open the following script from URL: https://raw.githubusercontent.com/amvmoody/hifi/master/unpublishedScripts/marketplace/audienceApplauseMeter/rezMeter.js
- Have a Microphone.
- Be sure to be in a domain with only the tester(s) present.

## Steps - Single Person
1. Verify all 3 pieces of the object load into the world.
2. Make some noise into your microphone and verify that the bar is moving from left to right in accordance to your loudness.
3. Let your noise level subside (mute mic) and verify the bar decreasing in size.
4. Verify that as the bar moves forward (Increase in loudness) the color is shifting from black to red.
5. Verify that as the bar moves backwards (decrease in loudness) the color is shifting from red to black.

## Steps - Group

1. Have everyone make a constant and relatively same level noise, (Clapping, whistling etc) verify the meter is moving. [Take note of the groups average loudness]
2. Have one person mute their microphone and record the average loudness of the group.
3. repeat step 2 until there are no more microphones left to mute. [record the results].
4. Compare the results and look to see if the initial average loudness is lowering in proportion to the number of microphones muted. ( assuming the average loudness of a group of 4 was initially 100%, the rest out the loudness' should read 75% with 1 muted microphone, 50% with 2 muted microphones, 25% with 3 muted microphones and 0% with all microphones muted.)

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
An admin must verify this patch for the pull request build to fire. Type "ok to test" to begin testing this pull request or "add to whitelist" to test this and all future pull requests from this collaborator.

**Comment by themelissabrown** [comment]
add to whitelist

**Comment by hifi-gustavo** [comment]
Build console output for commit: 29d086d3909856f1927a3e031c5e6c80b4890d45 available at: https://gist.github.com/1220c90c84396999fdee31617a41d6ef<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10232-29d086d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10232-29d086d.dmg

**Comment by ImRichardJackson** [comment]
Windows 10 - Single tester

QA fail- objects do not load in.

**Comment by amvmoody** [comment]
@ImRichardJackson It appears I linked the wrong url link in the QA description. This has been fixed :) https://raw.githubusercontent.com/amvmoody/hifi/master/unpublishedScripts/marketplace/audienceApplauseMeter/rezMeter.js

**Comment by ImRichardJackson** [comment]
hifi-snap-by-im_rj-on-2017-04-17_17-32-45

@amvmoody Awsome :D Here is what I was able to load in, are these the three object I should be seeing?

**Comment by amvmoody** [comment]
Yes everything appears there. You have the background (grey), the parent. then the Text entity(text), a child. the bar, (black to red), a child.

**Comment by ImRichardJackson** [comment]
Test 2 was a fail.

When speaking into the mic the objects were stagnant.

**Comment by amvmoody** [comment]
@ImRichardJackson In the Audio menu check the "show level meter" check if that is moving (upper left) as you speak, or that the mic is not muted. I cannot reproduce the no audio.

**Comment by ImRichardJackson** [comment]
Yes my mic/audio meter is moving.

**Comment by amvmoody** [comment]
@ImRichardJackson It was tested by another group last night and everything works as expected. Let me know a time you are in world (I'll be around 4PM PST onward tonight) so I can come see whats going on, see if we can get to the bottom of this mystery. :)

**Comment by ImRichardJackson** [comment]
Ok I'll be on around 4 pm as well.

**Comment by ImRichardJackson** [comment]
Passed with flying colors  :)

**Comment by themelissabrown** [comment]
Do not merge. Assets will be added to the Marketplace

File: unpublishedScripts/marketplace/audienceApplauseMeter/bakedTextMeter.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+        function average(a) {
+
+            var sum = 0;
+            var total = a.length;
+            for (var i = 0; i < total; i++) {
+                sum += a[i];
+            }
+
+            return Math.round(sum / total)
```
Comment (ZappoMan) [inline unpublishedScripts/marketplace/audienceApplauseMeter/bakedTextMeter.js]: coding standard: end statements with a `;`

File: unpublishedScripts/marketplace/audienceApplauseMeter/meter.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+        function average(a) {
+
+            var sum = 0;
+            var total = a.length;
+            for (var i = 0; i < total; i++) {
+                sum += a[i];
+            }
+
+            return Math.round(sum / total)
```
Comment (ZappoMan) [inline unpublishedScripts/marketplace/audienceApplauseMeter/meter.js]: coding standard: end statements with a `;`

File: unpublishedScripts/marketplace/audienceApplauseMeter/meter.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        function normalizedAverage(a) {
+            a = a.map(function (v) {
+                return Math.round(( 100 / MAX_AUDIO_THRESHOLD ) * v)
+            });
+            return average(a);
+        }
+
+        var norm = normalizedAverage(framePool) ;
```
Comment (ZappoMan) [inline unpublishedScripts/marketplace/audienceApplauseMeter/meter.js]: remove extra space at end

File: unpublishedScripts/marketplace/audienceApplauseMeter/meter.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+
+        framePool.push(average(avatarLoudnessPool));
+        if (framePool.length >= REFERENCE_FRAME_COUNT) {
+            framePool.shift();
+        }
+
+        function normalizedAverage(a) {
+            a = a.map(function (v) {
+                return Math.round(( 100 / MAX_AUDIO_THRESHOLD ) * v)
```
Comment (ZappoMan) [inline unpublishedScripts/marketplace/audienceApplauseMeter/meter.js]: coding standard: end statements with a `;`

File: unpublishedScripts/marketplace/audienceApplauseMeter/meter.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+        Entities.editEntity(textID, {text: "Loudness: % " + norm});
+
+        var barProperties = Entities.getEntityProperties(barID);
+        JSON.stringify(barProperties) ;
+
+        var colorShift = 2.55 * norm; //shifiting the scale to 0 - 255
+        var xShift = norm / 100; // changing scale from 0-100 to 0-1
+        var normShift = xShift - .5; //shifitg scale form 0-1 to -.5 to .5
+        var halfShift = xShift / 2 ;
```
Comment (ZappoMan) [inline unpublishedScripts/marketplace/audienceApplauseMeter/meter.js]: remove extra space

File: unpublishedScripts/marketplace/audienceApplauseMeter/textEntityMeter.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+        function average(a) {
+
+            var sum = 0;
+            var total = a.length;
+            for (var i = 0; i < total; i++) {
+                sum += a[i];
+            }
+
+            return Math.round(sum / total)
```
Comment (ZappoMan) [inline unpublishedScripts/marketplace/audienceApplauseMeter/textEntityMeter.js]: coding standard: end statements with a `;`

File: unpublishedScripts/marketplace/audienceApplauseMeter/textEntityMeter.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+
+        framePool.push(average(avatarLoudnessPool));
+        if (framePool.length >= REFERENCE_FRAME_COUNT) {
+            framePool.shift();
+        }
+
+        function normalizedAverage(a) {
+            a = a.map(function (v) {
+                return Math.round(( 100 / MAX_AUDIO_THRESHOLD ) * v)
```
Comment (ZappoMan) [inline unpublishedScripts/marketplace/audienceApplauseMeter/textEntityMeter.js]: coding standard: end statements with a `;`

**Comment by hifi-gustavo** [comment]
Build console output for commit: d805f75d5575b77bb9729d2a5f45020bb4114c52 available at: https://gist.github.com/1220c90c84396999fdee31617a41d6ef<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10232-d805f75.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10232-d805f75.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 01855c0d9c63bd8ec1c1b5f3087a47c487e5c8bf available at: https://gist.github.com/1220c90c84396999fdee31617a41d6ef<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10232-01855c0.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10232-01855c0.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 55fe2f9dc014bf4de8e133f8fdfe016ce3ab1c2c available at: https://gist.github.com/1220c90c84396999fdee31617a41d6ef<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10232-55fe2f9.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10232-55fe2f9.dmg

**Comment by themelissabrown** [comment]
@ZappoMan can you check and clear conflicts so this can be approved and merged?

**Comment by themelissabrown** [comment]
updated assets need more coding. @amvmoody will let me know when it's ready for testing and code review again

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8daf9698bd2a048a3eafcb94b665f9c5735bbef0 available at: https://gist.github.com/1220c90c84396999fdee31617a41d6ef<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10232-8daf969.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10232-8daf969.dmg

**Comment by amvmoody** [comment]
# QA

## Info
- recommended to be tested with a group.


## Setup
 - Open the following script from URL: https://raw.githubusercontent.com/amvmoody/hifi/8daf9698bd2a048a3eafcb94b665f9c5735bbef0/unpublishedScripts/marketplace/audienceApplauseMeter/rezApplauseOmeter.js
- Have a Microphone.
- Be sure to be in a domain with only the tester(s) present.

## Steps - Single Person
1. Verify object loads into world.
2. Make some noise into your microphone and verify that the needle is moving from left to right in accordance to your loudness.
3. Let your noise level subside (mute mic) and verify the needle returns to the zero state (left most).


## Steps - Group

1. Have everyone make a constant and relatively same level noise, (Clapping, whistling etc) verify the meter is moving. [Take note of the groups average loudness]
2. Have one person mute their microphone and record the average loudness of the group.
3. repeat step 2 until there are no more microphones left to mute. [record the results].
4. Compare the results and look to see if the initial average loudness is lowering in proportion to the number of microphones muted. ( assuming the average loudness of a group of 4 was initially 100%, the rest out the loudness' should read 75% with 1 muted microphone, 50% with 2 muted microphones, 25% with 3 muted microphones and 0% with all microphones muted.)

**Comment by Nex-Pro** [comment]
Looking at this as per @themelissabrown

**Comment by themelissabrown** [comment]
@amvmoody This passed the second test with updated assets. Please confirm you're ready for me to remove the DNM label

**Comment by amvmoody** [comment]
@themelissabrown if you are going to use Github to store assets (images, fbx) then i see no reason to prevent this from being merged.

**Pull request merged**